gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+problem is to obtain a deterministic hash-chain of ''each'' filter. This hash
+chain or "filter header chain" is similar to the regular Bitcoin headers in
+that it allows a light client to verify the ''authenticity'' of a received
+filter.
+
+The filter header chain for a particular filter type is described by the
+following recurrence:
+<pre>
+filter_header(n: uint) -> [32]byte = 
+   let zero_hash [32]byte = {0*32}
+
+   if n == 0:
+       double-sha-256(genesis_block.prevblock || filter(0))
+
+   match filter(n):
+      case Some:
+          double-sha-256(filter_header(n-1) || double-sha-256(filter(n)))
+      case None:
+          double-sha-256(filter_header(n-1) || double-sha-256(zero_hash))
+
+   where filter(n) is the filter for block height n
+</pre>
+
+The filter header for the genesis block uses the hash stored in the prevblock
+field of the genesis block header itself, as there's no prior filter header
+(by definition).
+
+Due to the nature of filter construction, it's possible to construct a block
+such that an "empty" filter will be produced. This is the case of a coinbase
+transaction that has no data pushes in its public key script. In this case, the
+"hash" of said filter is simply "32 zeroes". 
+
+This filter header chain can be utilized by light clients to gain a greater
+degree of security against bamboozling full-nodes during their initial chain
+sync. In addition to fetching all the bitcoin headers, light clients
+implementing this BIP should also fetch all the ''filter headers'' from ''each'' of
+their connected peers. This allows light clients to efficiently detect nodes
+that advertise a conflicting filter chain history (by ensuring all nodes return
+the ''same'' filter header hash for a particular block hash).
+
+Additionally, the filter header chain also allows a light client to efficiently
+verify purported filter authenticity when fetching the next set of headers from
+chain tip. Instead of fetching the filter ''from each peer'' (which wastes
+bandwidth), a light client instead does the following:
+
+<pre>
+verify_from_tip(tip_block_hash: [32]byte):
+    let filter_types = {supported_filter_types...}
+    let connected_peers = {list_of_connected_full_nodes...}
+
+    for filter_type in filter_types:
+
+        let filter_headers = set()
+        for peer in connected_peers:
+            let filter_header = peer.fetch_filter_header(tip_block_hash)
+            filter_headers.insert(filter_header)
+
+        if len(filter_headers) != 1:
+            // Peers have conflicting filters. The light client should fetch
+            // each unique filter from the set of peers AND fetch the block. The
+            // light client can then verify which filter header is correct, and
+            // BAN the offending peers.
+
+        // Otherwise, syncing continues as normal: fetch filter to see if it
+        // matches any relevant items.
+</pre>
+
+Light clients should persistently commit all filter headers to disk, as when
```

Comment:
I dont think this should be normative, but if it is, I'd suggest using RFC words and use MAY here instead of SHOULD.

Reply:
Modified to use MAY.